operator/internal/configwatcher/configwatcher.go

```diff
@@ -0,0 +1,242 @@
+	return func(c *ConfigWatcher) {
+		c.configPath = path
+	}
+}
+
+func WithUsersDirectory(path string) Option {
+	return func(c *ConfigWatcher) {
+		c.usersDirectory = path
+	}
+}
+
+func WithFs(fs afero.Fs) Option {
+	return func(c *ConfigWatcher) {
+		c.fs = fs
+	}
+}
+
+func WithInitializedSignal(ch chan struct{}) Option {
+	return func(c *ConfigWatcher) {
+		c.initialized = ch
+	}
+}
+
+func NewConfigWatcher(log logr.Logger, watch bool, options ...Option) *ConfigWatcher {
+	watcher := &ConfigWatcher{
+		log:            log,
+		watch:          watch,
+		configPath:     defaultConfigPath,
+		usersDirectory: defaultUsersDirectory,
+		fs:             afero.NewOsFs(),
+		initialized:    make(chan struct{}),
+	}
+
+	for _, option := range options {
+		option(watcher)
+	}
+
+	return watcher
+}
+
+func (w *ConfigWatcher) Start(ctx context.Context) error {
+	params := rpkconfig.Params{ConfigFlag: w.configPath}
+
+	config, err := params.Load(w.fs)
+	if err != nil {
+		return fmt.Errorf("loading rpk config: %w", err)
+	}
+
+	factory := internalclient.NewFactory(&rest.Config{}, nil).WithFS(w.fs)
+	client, err := factory.RedpandaAdminClient(ctx, config.VirtualProfile())
+	if err != nil {
+		return fmt.Errorf("initializing Redpanda admin API client: %w", err)
+	}
+
+	w.adminClient = client
+
+	close(w.initialized)
+
+	w.syncInitial(ctx)
+
+	return w.watchFilesystem(ctx)
+}
+
+func (w *ConfigWatcher) syncInitial(ctx context.Context) {
+	files, err := os.ReadDir(w.usersDirectory)
+	if err != nil {
+		w.log.Error(err, "unable to get user directory files")
+		return
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		filePath := path.Join(w.usersDirectory, file.Name())
+		w.SyncUsers(ctx, filePath)
+	}
+}
+
+func (w *ConfigWatcher) watchFilesystem(ctx context.Context) error {
+	if !w.watch {
+		<-ctx.Done()
+		return nil
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	if err := watcher.Add(w.usersDirectory); err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case err := <-watcher.Errors:
+			// here we don't return as that'd crash the broker, instead
+			// just log the error and move on after some sleep time.
+			w.log.Error(err, "watcher returned an error")
+			time.Sleep(5 * time.Second)
+		case event := <-watcher.Events:
+			file := path.Join(w.usersDirectory, event.Name)
+			w.SyncUsers(ctx, file)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (w *ConfigWatcher) SyncUsers(ctx context.Context, path string) {
+	file, err := w.fs.Open(path)
+	if err != nil {
+		w.log.Error(err, "unable to open superusers file", "file", path)
+		return
+	}
+	defer file.Close()
+
+	// sync our internal superuser first
+	internalSuperuser, password, mechanism := getInternalUser()
+	// the internal user should only ever be created once, so don't
+	// update its password ever.
+	w.syncUser(ctx, internalSuperuser, password, mechanism, false)
```

Comment:
This is not true as `SuncUsers` function is called from 2 places `syncInitial` and `watchFilesystem`.

Reply:
But the last argument "recreate" does the trick!

Reply:
NIT: Maybe comment could be adjusted to indicate that syncUser will not recreate user as the last flag is set to `false`.